shell/platform/glfw/flutter_glfw.cc

```diff
@@ -690,6 +693,123 @@ static bool RunFlutterEngine(
+}
+
+const char* GetLocaleStringFromEnvironment() {
+  return GetLocaleStringFromEnvironmentVariables(
+      getenv("LANGUAGE"), getenv("LC_ALL"), getenv("LC_MESSAGES"),
+      getenv("LANG"));
+}
+
+// Parse a locale into its components.
+void ParseLocale(const std::string& locale,
+                 std::string* language,
+                 std::string* territory,
+                 std::string* codeset,
+                 std::string* modifier) {
+  // Locales are in the form "language[_territory][.codeset][@modifier]"
+  std::string::size_type end = locale.size();
+  std::string::size_type modifier_pos = locale.rfind('@');
+  if (modifier_pos != std::string::npos) {
+    *modifier = locale.substr(modifier_pos + 1, end - modifier_pos - 1);
+    end = modifier_pos;
+  }
+
+  std::string::size_type codeset_pos = locale.rfind('.');
+  if (codeset_pos != std::string::npos) {
+    *codeset = locale.substr(codeset_pos + 1, end - codeset_pos - 1);
+    end = codeset_pos;
+  }
+
+  std::string::size_type territory_pos = locale.rfind('_');
+  if (territory_pos != std::string::npos) {
+    *territory = locale.substr(territory_pos + 1, end - territory_pos - 1);
+    end = territory_pos;
+  }
+
+  *language = locale.substr(0, end);
+}
+
+std::vector<std::unique_ptr<FlutterLocale>> GetLocales(
+    const char* locale_string,
+    std::list<std::string>& locale_storage) {
+  if (!locale_string || locale_string[0] == '\0') {
+    locale_string = "C";
+  }
+  std::istringstream locales_stream(locale_string);
+  std::vector<std::unique_ptr<FlutterLocale>> locales;
+  std::string s;
+  while (getline(locales_stream, s, ':')) {
+    locales.push_back(std::make_unique<FlutterLocale>());
+    std::string language, territory, codeset, modifier;
+    ParseLocale(s, &language, &territory, &codeset, &modifier);
+    FlutterLocale* locale = locales.back().get();
+    locale->struct_size = sizeof(FlutterLocale);
+    if (!language.empty()) {
+      locale_storage.push_back(language);
+      locale->language_code = locale_storage.back().c_str();
+    }
+    if (!territory.empty()) {
+      locale_storage.push_back(territory);
+      locale->country_code = locale_storage.back().c_str();
+    }
+    if (!codeset.empty()) {
+      locale_storage.push_back(codeset);
+      locale->script_code = locale_storage.back().c_str();
+    }
+    if (!modifier.empty()) {
+      locale_storage.push_back(modifier);
+      locale->variant_code = locale_storage.back().c_str();
+    }
+  }
+  return locales;
+}
+
+// Returns parsed locales fetched from the environment.
+std::vector<std::unique_ptr<FlutterLocale>> GetLocalesFromEnvironment(
+    std::list<std::string>& locale_storage) {
+  return GetLocales(GetLocaleStringFromEnvironment(), locale_storage);
+}
+
+// Passes locale information to the Flutter engine.
+static void SetUpLocales(FlutterDesktopEngineState* state) {
+  // Helper to extend lifetime of the strings passed to Flutter.
+  std::list<std::string> locale_storage;
```

Comment:
This construction seems more confusing than just inlining the conversion to C strings (as we do on other platforms IIRC).

Reply:
You need something to extend the lifetimes based on the calling convention, and this is basically how the other linux shell does it. I didn't think to look at the windows code, that is a much better way to handle it, I will mirror that over here.